File: 16-3sum-closest/16-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums_sort = sorted(nums)
        def findTwo(first, index):
            left, right = index+1, len(nums_sort) - 1
            diff = 10**5
            while(left < right):
                second, third = nums_sort[left], nums_sort[right]
                cur = first + second + third
                diff_cur = abs(target-cur)
                if cur == target:
                    return target
                if diff_cur < diff:
                    diff = diff_cur
                    res = cur 
                if left+1 == right:
                    return res
                if cur < target:
                    left = left + 1
                else:
                    right = right - 1
        n = len(nums)
        diff = 10**5
        for i in range(n):
            if i <= n - 3:
                cur = findTwo(nums_sort[i],i)
                diff_cur = abs(target-cur)
                if cur == target:
                    return target
                if diff_cur < diff:
                    diff = diff_cur
                    res = cur
        return res

File: 16-3sum-closest/16-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int N = nums.length;
        int ans = 300000;
        for(int i = 0; i < N - 2; i++){
            int cur_num = nums[i];
            int newTarget = target - cur_num;
            int cur_res = this.twoSumClosest(i + 1, nums, newTarget);
            if (Math.abs(cur_res + cur_num - target) < Math.abs(ans - target)){
                ans = cur_res + cur_num;
            }
            if (ans == target) return target;
                
        }
        return ans;
        
    }
    public int twoSumClosest(int start, int[] nums, int target){
        int N = nums.length;
        int left = start, right = N - 1;
        int res = 300000;
        while(left < right){
            int cur_sum = nums[left] + nums[right];
            if (Math.abs(res - target) > Math.abs(cur_sum - target)){
                res = cur_sum;
            }
            if (cur_sum == target) return target;
            if (cur_sum < target)
                left++;
            if (cur_sum > target)
                right--;
            
        }
        return res;
        
    }
}